tests/jballerina-unit-test/src/test/resources/test-src/query/match-stmt-in-do-clause.bal

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2023 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+function testConstMatchPattern() {
+    string[][] data = [
+        [ "abc", "def", "ghi"],
+        [ "jkl", "mno"],
+        [ "pqr", "stu", "vwx", "yz"],
+        ["123"]
+    ];
+
+    (string|int)[] result = [];
+    from string[] row in data
```

Comment:
Shall we have a test where we have  an array of union type for `data` and `row` also have an union type?

Reply:
Basically we need to check if type narrowing works with query expressions

Reply:
Added a test